python_modules/dagster/dagster/_core/definitions/external_asset_graph.py

```diff
@@ -122,106 +93,214 @@ def from_repository_handles_and_external_asset_nodes(
         repo_handle_external_asset_nodes: Sequence[Tuple[RepositoryHandle, "ExternalAssetNode"]],
         external_asset_checks: Sequence["ExternalAssetCheck"],
     ) -> "ExternalAssetGraph":
-        upstream: Dict[AssetKey, AbstractSet[AssetKey]] = {}
-        source_asset_keys: Set[AssetKey] = set()
-        partitions_defs_by_key: Dict[AssetKey, Optional[PartitionsDefinition]] = {}
-        partition_mappings_by_key: Dict[AssetKey, Dict[AssetKey, PartitionMapping]] = defaultdict(
-            defaultdict
-        )
-        group_names_by_key = {}
-        freshness_policies_by_key = {}
-        auto_materialize_policies_by_key = {}
-        backfill_policies_by_key = {}
-        keys_by_atomic_execution_unit_id: Dict[str, Set[AssetKeyOrCheckKey]] = defaultdict(set)
         repo_handles_by_key = {
             node.asset_key: repo_handle
             for repo_handle, node in repo_handle_external_asset_nodes
             if not node.is_source or node.is_observable
         }
-        job_names_by_key = {
-            node.asset_key: node.job_names
-            for _, node in repo_handle_external_asset_nodes
-            if not node.is_source or node.is_observable
+
+        non_source_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
+        observable_source_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
+        non_observable_source_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
```

Comment:
we need to rename `ExternalAssetNode` or else we might go crazy.

Reply:
Ok given this we need clearer names and copious comments here. Are `observable_source_node_pairs` just assets represented by `SourceAsset`, or is it inclusive of external observables?

Reply:
"source" here in general is inclusive of observables-- when the `ExternalAssetNode` list is built in the code server external assets and source assets both have `is_source` set to `True`. I've changed the names and added a comment.